grudge/op.py

```diff
@@ -403,155 +998,147 @@ def norm(dcoll, vec, p, dd=None):
         else:
             raise ValueError("unsupported norm order")
 
-    return _norm(dcoll, p, dd)(arg=vec)
-
-
-@memoize_on_first_arg
-def _nodal_reduction(dcoll, operator, dd):
-    return bind(dcoll, operator(dd)(sym.var("arg")), local_only=True)
+    return _norm(dcoll, vec, p, dd)
 
 
 def nodal_sum(dcoll, dd, vec):
-    return _nodal_reduction(dcoll, sym.NodalSum, dd)(arg=vec)
+    from warnings import warn
+    warn("Using 'nodal_sum' is deprecated, use 'nodal_summation' instead.",
+         DeprecationWarning, stacklevel=2)
+    return nodal_summation(vec)
 
 
-def nodal_min(dcoll, dd, vec):
-    return _nodal_reduction(dcoll, sym.NodalMin, dd)(arg=vec)
-
-
-def nodal_max(dcoll, dd, vec):
-    return _nodal_reduction(dcoll, sym.NodalMax, dd)(arg=vec)
+def nodal_summation(vec):
+    r"""Return the nodal sum of a vector of degrees of freedom *vec*.
 
-# }}}
+    :arg vec: a :class:`~meshmode.dof_array.DOFArray`.
+    :returns: an integer denoting the nodal sum.
+    """
+    actx = vec.array_context
+    return sum([actx.np.sum(grp_ary) for grp_ary in vec])
```

Comment:
@inducer I'm not rethinking this choice (similarly for `op.nodal_{minimum, maximum}`). I think we will probably want to keep `dcoll` as a passed argument. The reason being is that I'm thinking ahead for MPI-aware nodal reductions (https://github.com/inducer/grudge/issues/16) and we'll definitely need the discretization collection for doing cross-rank reductions. What do you think?

Reply:
I agree with keeping `dcoll` as an argument.